anywidget/experimental.py

```diff
@@ -103,3 +106,31 @@ def _decorator(cls: T) -> T:
         return cls
 
     return _decorator(cls) if cls is not None else _decorator  # type: ignore
+
+
+def _register_experimental_custom_message_reducer(
+    widget: AnywidgetReducerProtocol,
+) -> None:
+    """Register a custom message reducer for a widget if it implements the protocol."""
+    # Only add the reducer if it doesn't already exist
+    if not hasattr(widget, "_experimental_anywidget_reducer"):
+        return
+
+    def handle_anywidget_dispatch(
+        self: AnywidgetReducerProtocol, msg: str | list | dict, buffers: list[bytes]
+    ) -> None:
+        if not isinstance(msg, dict) or msg.get("kind") != "anywidget-dispatch":
+            return
+        response, buffers = widget._experimental_anywidget_reducer(
+            msg["action"], buffers
+        )
```

Comment:
I'd suggest for this to be in a try/except. I think if an error is raised in the user callback, then the error gets swallowed somewhere. Maybe on error you could propagate something back to the JS so you're not blocking the Python kernel any more?

Reply:
Ah, that's a great idea! We could throw a JS error for a special message sent back from Python.